### PR TITLE
fix(Text) Fix style transfer issue on a line that is not empty

### LINKED ISSUE #9028: [BUG]: text style onInput
@asturur please help if you can

**It seems that when text is wrapped styles is not updated correctly from onInput**

I have to say, I do understand why styles was chosen to be a map but I don't think that was a good choice.
So much easier to maintain and update it as a flat array corresponding to char offset.
Then wrapping is not related and syncing char changes from `onInput` to styles or where ever is as simple as splicing 
I do think that fixing this bug or making it an array will cost the same amount of time.
I understand it is breaking so you will probably not want this now.
The logic for me is unapproachable.

Suspected `insertNewlineStyleObject` and `removeStyleFromTo`:

https://github.com/fabricjs/fabric.js/assets/34343793/35b086e7-7ae3-4858-b471-01a4bb91b890

https://github.com/fabricjs/fabric.js/assets/34343793/a66cc38a-b33e-4a5f-9b51-18d6c70abb5a

### PR DESCRIPTION
**Pull request by asturur**
## Motivation

Fix las bug mentioned in #9028 that adds unwanted style on previous text when we insert a new line.
This bug is also present in 5.x

image

bug after inserting a new line

image

 
close #9028

## Changes

- when insert a new line and some text is available after the newLine we won't prestyle the char 0 of the new created line.

### CODE REVIEW

**Comment by asturur** [comment]
This in theory removes the last known style text bug with typing.
I'm sure that with some combination of copy paste we can trigger some uncovered case

File: e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png
Comment (ShaMan123) [inline e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: there is another major bug here
fabric.js should be in the first line and not break, this is a regression I am seeing at work as well

**Review by ShaMan123 — COMMENTED** [review]
I really do not like these patches
As I said and PRed the problem is deeper so you will keep finding these annoying edge cases
Reviewing this will take time

File: e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png
Comment (asturur) [reply e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: this is splitByGrapheme true so this will not take care of words, will just wraps whenever it feels it has to wrap.

**Comment by asturur** [comment]
> I really do not like these patches As I said and PRed the problem is deeper so you will keep finding these annoying edge cases Reviewing this will take time

As i said, we don't refactor to fix bugs, and moreover we are not refactoring text anytime soon.
If we want to improve text code we need to spin up a new class that will slowly replace the older, so small dumb fixes to the current text code is all we can do.

**Comment by asturur** [comment]
Anyway there is no rush here to merge this, this particular bug is here since before 5.

File: src/shapes/IText/ITextBehavior.ts [hunk 1]
```diff
@@ -805,9 +805,10 @@ export abstract class ITextBehavior<
     copiedStyle?: { [index: number]: TextStyleDeclaration }
   ) {
     const newLineStyles: { [index: number]: TextStyleDeclaration } = {};
-    const isEndOfLine =
-      this._unwrappedTextLines[lineIndex].length === charIndex;
-    let somethingAdded = false;
+    const originalLineLength = this._unwrappedTextLines[lineIndex].length;
+    const isEndOfLine = originalLineLength === charIndex;
+
+    let someStyleIsCarringOver = false;
```
Comment (gloriousjob) [inline src/shapes/IText/ITextBehavior.ts:811]: Do you mean someStyleIsCarryingOver?

File: src/shapes/IText/ITextBehavior.ts [hunk 2]
```diff
@@ -828,14 +829,16 @@ export abstract class ITextBehavior<
...
       // if is end of line, the extra style we copied
       // is probably not something we want
       this.styles[lineIndex + qty] = newLineStyles;
       styleCarriedOver = true;
     }
-    if (styleCarriedOver) {
+    if (styleCarriedOver || originalLineLength > charIndex) {
       // skip the last line of since we already prepared it.
+      // or contains text without style that we don't want to style
+      // just becuase it changed line
```
Comment (gloriousjob) [inline src/shapes/IText/ITextBehavior.ts:841]: because

**Comment by gloriousjob** [comment]
Just checking but the example shows splitting a styled, then unstyled into two lines. Is the reverse handled as well (i.e., foo is unstyled, bar is red, and splitting doesn't make the b unstyled)?

File: src/shapes/IText/ITextBehavior.ts (hunk 1 above)
Comment (ShaMan123) [reply src/shapes/IText/ITextBehavior.ts:811]: @gloriousjob you can use the suggestion markdown
image

```suggestion
    let someStyleIsCarryingOver = false;
```

File: e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png
Comment (ShaMan123) [reply e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: If you look at v5 that is not the same I think

File: e2e/tests/text/adding-text/index.spec.ts [hunk 3]
```diff
@@ -80,5 +81,29 @@ for (const splitByGrapheme of [true, false]) {
...
+    await canvasUtil.click({
+      position: clickPointEnd,
+      delay: 200,
+    });
+    await page.mouse.down();
+    await page.mouse.move(1, clickPointEnd.y + 150, { steps: 15 });
+    await page.mouse.up();
+    // we remove them because space is finishing
+    await canvasUtil.press('Backspace');
+    // lets click where style end to show that we can add new line without carrying over
```
Comment (ShaMan123) [inline e2e/tests/text/adding-text/index.spec.ts:95]: I suggest usign `test.step` to organize the test instead of leaving comments
It is more clear when reading and when looking at test results

**Review by ShaMan123 — CHANGES_REQUESTED** [review]
Playwright is a very very good tool that makes reviewing this PR simple
However the test passes on current master even though it is broken, please make it break

File: e2e/tests/text/adding-text/index.spec.ts (hunk 3 above)
Comment (ShaMan123) [reply e2e/tests/text/adding-text/index.spec.ts:95]: screenshot
[85800a532aaae57a102d53213c2b8114a0cadb8e.webm](https://github.com/fabricjs/fabric.js/assets/34343793/dfb8a1cc-cd8a-4f35-9db6-db0b9a93557a)

Comment (ShaMan123) [reply e2e/tests/text/adding-text/index.spec.ts:95]: image

File: e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png
Comment (gloriousjob) [reply e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: Setting split should split the word up (as in the other split true tests). I wondered about the word grapheme since this is really a Unicode character split but apparently there's different takes on what's considered a character vs. grapheme. I think the question here is what was broken in v5 that was considered "working" and what is actually a regression.

File: src/shapes/IText/ITextBehavior.ts [hunk 4]
```diff
@@ -828,14 +829,16 @@ export abstract class ITextBehavior<
       }
     }
     let styleCarriedOver = false;
-    if (somethingAdded && !isEndOfLine) {
+    if (someStyleIsCarringOver && !isEndOfLine) {
```
Comment (gloriousjob) [inline src/shapes/IText/ITextBehavior.ts:832]:
```suggestion
    if (someStyleIsCarryingOver && !isEndOfLine) {
```

File: src/shapes/IText/ITextBehavior.ts (hunk 2 above)
Comment (gloriousjob) [reply src/shapes/IText/ITextBehavior.ts:841]:
```suggestion
      // just because it changed lines
```

File: e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png
Comment (asturur) [reply e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: splitByGrapheme is an option that was introduced for languages that do not use spaces to break text.
It doesn't split by unicode character but by grapheme clusters, if the appropriate splitter is provided.
Otherwise it splits with the rudimental grapheme splitter of fabricJS.
a grapheme can be made of more unicode codepoints/characters glued together. In our example it doesn't happen.

Comment (asturur) [reply e2e/tests/text/adding-text/index.spec.ts-snapshots/5-after-pasting-splitByGrapheme-true.png:1]: regarding ' I think the question here is what was broken in v5 that was considered "working" and what is actually a regression.' i think nothing broken in v5 was considered working for styles.
This bug wasn't listed anywhere and i didn't know or remember of it.

**Comment by asturur** [comment]
i will add the test in master first, with steps as commented.
https://github.com/fabricjs/fabric.js/pull/9531/files?diff=unified&w=1

We are adding an extra test step to cover the bug, and we are adding a step construct between the blocks of the test.
The screenshots are slightly smaller to avoid dealing with too small percentages

**Comment by asturur** [comment]
hopefully test fail, but i wouldn't count on it.

**Comment by asturur** [comment]
so at commit [29dcc2e](https://github.com/fabricjs/fabric.js/pull/9461/commits/29dcc2e88261cc059d45f2691f510de48b857a3c) we have failure

**Comment by asturur** [comment]
and at commit [b84e6f8](https://github.com/fabricjs/fabric.js/pull/9461/commits/b84e6f87694a1abfdee671cd7810e2b55e62e87f) we have good test.

Locally it passes anyway, but i can't do miracles, i should redo a different test that show this particular but bigger or find a way to only have a zone of the screenshot, but i can't find an option for it.

But also this is good enough

File: e2e/tests/text/adding-text/index.spec.ts [hunk 5]
```diff
@@ -112,7 +112,7 @@ for (const splitByGrapheme of [true, false]) {
       // and there is no style on the part of text that follows, but there is visible text.
       await expect(await canvasUtil.screenshot()).toMatchSnapshot({
         name: `6-after-adding-a-newline-splitByGrapheme-${splitByGrapheme}.png`,
-        maxDiffPixelRatio: 0.01,
+        maxDiffPixelRatio: 0.009,
```
Comment (ShaMan123) [inline e2e/tests/text/adding-text/index.spec.ts:115]:
```
await expect(page).toHaveScreenshot(`${name}.png`, {
          clip: {
            x,
            y,
            width,
            height,
          },
          maxDiffPixels, // insteadof percentage
        });
```

**Review by ShaMan123 — COMMENTED** [review]
There a typos that are commented by @gloriousjob 
And I cpmmented a way tp clip the snapshot and perform a more reliable diff using number of pixels instead of percentage

File: e2e/tests/text/adding-text/index.spec.ts (hunk 5 above)
Comment (asturur) [reply e2e/tests/text/adding-text/index.spec.ts:115]: where is this clip in the documentation? i looked for something like this.

Comment (asturur) [reply e2e/tests/text/adding-text/index.spec.ts:115]: oh ok is an entirely different way to do visual comparison. the locator can't be clipped.

Comment (ShaMan123) [reply e2e/tests/text/adding-text/index.spec.ts]: but the image can, right?

File: e2e/tests/text/adding-text/index.spec.ts [hunk 6]
```diff
@@ -110,10 +110,18 @@ for (const splitByGrapheme of [true, false]) {
...
+      await expect(page).toHaveScreenshot(
+        `6-after-adding-a-newline-splitByGrapheme-${splitByGrapheme}.png`,
+        {
+          clip: {
+            x: 0,
+            y: clickPointYellow.y - 20,
+            width: 120,
+            height: 120,
+          },
+          maxDiffPixels: 0.029,
```
Comment (ShaMan123) [inline e2e/tests/text/adding-text/index.spec.ts:122]: I thought this was an interger

**Review by ShaMan123 — COMMENTED** [review]
The playwright test is failing

**Review by ShaMan123 — APPROVED** [review]
